README.md

```diff
@@ -1,16 +1,658 @@
+
+All of the roles from the referenced UCANs MUST persist to the invocation per the [roles table](#2-roles).
+
+The invocation wrapper MUST be signed by the same principal that issued the UCAN.
+
+## 3.1 Fields
+
+| Field         | Type                                         | Value     | Description                                      | Required | Default |
+|---------------|----------------------------------------------|-----------|--------------------------------------------------|----------|---------|
+| `ucan/invoke` | `CID`                                        |           | The CID of the UCAN to invoke                    | Yes      |         |
+| `v`           | `SemVer`                                     | `"0.1.0"` | SemVer of the UCAN invocation this schema        | Yes      |         |
+| `run`         | `"*" or {String: {URI : {Ability : [Any]}}}` |           | Which UCAN capabilities to run                   | No       | `"*"`   |
+| `nnc`         | `String`                                     |           | A unique nonce, to distinguish each invocation   | Yes      |         |
+| `ext`         | `Any`                                        |           | Non-normative extended fields                    | No       | `null`  |
+| `sig`         | `Bytes`                                      |           | Signature of the rest of the field canonicalized | Yes      |         |
+  
+### 3.1.1 Invocation
+
+The `ucan/invoke` field MUST contain CIDs pointing to the UCANs to invoke. The outmost UCAN being invoked SHOULD NOT contain any actions that are not intended to be executed.
+
+### 3.1.2 Version
+
+The `v` field MUST contain the version of the invocation object  schema.
+
+### 3.1.3 Run Capabilities
+
+The OPTIONAL `run` field MUST reference the actions contained in the UCAN are to be run during the invocation. To run all actions in the underlying UCAN, the `"*"` value MUST be used. If only specific actions (or [pipelines](#5-promise-pipelining)) are intended to be run, then they MUST be prefixed with an arbitrary label and treated as a UCAN attenuation: all actions MUST be backed by a matching capability of equal or lesser scope.
+
+#### 3.1.3.1 Promises
+
+The only difference from general capabilities is that [promises](#5-promise-pipelining) MAY also be used as inputs to attenuated fields.
+
+If a capability input has the key `"_"` and the value is a promise, the input MUST be discarded and only used for determining sequencing actions.
+
+### 3.1.4 Nonce
+
+The `nnc` field MUS contain a unique nonce. This field makes each invocation unique.
+
+### 3.1.5 Extended Fields
+
+The OPTIONAL `ext` field MAY contain arbitrary data. If not present, the `ext` field MUST me interpreted as `null`, including for [signature](#315-signature).
+
+### 3.1.6 Signature
+
+The `sig` field MUST contain the signature of the other fields. To construct the payload, the other fields MUST first be canonicalized as `dag-cbor`.
```

Comment:
A couple of thoughts:
- IIUC, this means we have to encode all but the `sig` field twice. I.e. we first need to dag-cbor encode an object with all except the `sig` field, then construct another object with `sig` added and re-encode, right? If instead we'd nest the signed fields into a `payload: Bytes` field or perhaps even just `payload: <object>`, then we'd save some encoding work.
- Do we want to assume dag-cbor in this spec? I guess this wouldn't benefit from the JWT format? And we're not trying to swim with the masses here and take advantage of a huge JSON ecosystem, unlike what we're doing in the UCAN spec?

Reply:
> Do we want to assume dag-cbor in this spec? I guess this wouldn't benefit from the JWT format? And we're not trying to swim with the masses here and take advantage of a huge JSON ecosystem, unlike what we're doing in the UCAN spec?

It's a good question! We certainly don't need JWT because, well, it's not an auth token.

>  If instead we'd nest the signed fields into a payload: Bytes field or perhaps even just payload: <object>, then we'd save some encoding work.

It's worth thinking about! I don't know if wrapping and unwrapping is going to make a huge difference vs removing one field on validation (since we don't have it yet on signing in either case).

```js
{
 "ucan/invoke": ["bafkreifzjut3te2nhyekklss27nh3k72ysco7y32koao5eei66wof36n5e"],
 "v": "0.1.0",
 "nnc": "abcdef",
 "run": "*", // Explicitly "run all"
 "ext": null,
 "sig": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt-VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"
}

// VS

{
  "ucan/invoke": {
    "v": "0.1.0",
    "prf": ["bafkreifzjut3te2nhyekklss27nh3k72ysco7y32koao5eei66wof36n5e"],
    "nnc": "abcdef",
    "run": "*", // Explicitly "run all"
    "ext": null,
  }
  "sig": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt-VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"
}
```

I guess it lets us move the `ucan/invoke` up one level, which is nice.

Reply:
> I don't know if wrapping and unwrapping is going to make a huge difference vs removing one field on validation

The cool thing about CBOR is that, if you have an object of this format:
```json
{
  "ucan/invoke": {
    "v": "0.1.0",
    "prf": ["bafkreifzjut3te2nhyekklss27nh3k72ysco7y32koao5eei66wof36n5e"],
    "nnc": "abcdef",
    "run": "*",
    "ext": null,
  },
  "sig": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt-VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"
}
```
And you serialize the object under `ucan/invoke` into bytes, then you've done most of the work for serializing the whole object into bytes. You simply need to take them and attach some information before and after it.
So there's no wrapping/unwrapping work, I think!

Reply:
We've possibly moving the encoding info to Varsig, but this actually reads perfectly well so made the change. Thanks @matheus23!